lib/iris/tests/integration/experimental/test_regrid_ProjectedUnstructured.py

```diff
@@ -68,16 +68,23 @@ def test_nearest_gnomonic_uk_domain(self):
         res = self.src.regrid(uk_grid, ProjectedUnstructuredNearest(crs))
 
         self.assertArrayShapeStats(
-            res, (1, 6, 17, 11), 315.8873266, 11.0006664668, rtol=1e-8
+            res,
+            (1, 6, 17, 11),
+            315.8854720963427,
+            11.000539210625737,
+            rtol=1e-8,
         )
         self.assertArrayShapeStats(
-            res[:, 0], (1, 17, 11), 299.99993826, 4.1356150388e-5
+            res[:, 0],
+            (1, 17, 11),
+            299.9999985207442,
+            3.53574517015874e-05,
         )
         expected_subset = np.array(
             [
-                [318.936829, 318.936829, 318.936829],
-                [318.936829, 318.936829, 318.936829],
-                [318.935163, 318.935163, 318.935163],
+                [318.92881733, 318.92881733, 318.92881733],
+                [318.92881733, 318.92881733, 318.92881733],
+                [318.92881733, 318.92881733, 318.92881733],
```

Comment:
I can't identify the root cause for the change in numbers here... hence why I've introduced a minimum pin for `cartopy` in `iris` (this is a major release update of `proj`)

We either accept this change "as is"... or dig deeper.

Reply:
What is the provenance of the original numbers in the test?  Assuming these are degrees longitude the difference is 0.00635deg ~ 700m on the equator.

Do we at least understand why the original test array was 2 rows of 318.936829 and 1 row of 318.935163, while the replacement is 3 identical rows of 318.92881733?

Reply:
> What is the provenance of the original numbers in the test? Assuming these are degrees longitude the difference is 0.00635deg ~ 700m on the equator.

I don't _think_ there is any significance to the values being tested in the array, other than a random snapshot. @pp-mo Any ideas on the history of this one?

I _think_ the philosophy of the test is rather weak, in that it's attempting to give "some level" of confidence that nothing has changed... but when something does change (like now) there is no context to determine whether the change is significant and/or valid.

I'm more inclined to trust a plot rather than raw numbers for this type of integration test :thinking:

Reply:
> I think the philosophy of the test is rather weak, in that it's attempting to give "some level" of confidence that nothing has changed... but when something does change (like now) there is no context to determine whether the change is significant and/or valid.

:+1: :grin:


> @pp-mo Any ideas on the history of this one?

I don't think there is anything special here, just a desire to check a small section of individual points, not just statistics over the whole array.

An actual change equivalent to 100's of metres does seem rather large though.  Maybe the default ellipsoid for this changed ?
I **_would_** have suggested that we should apply a more sensible tolerance to AssertArrayEqual.
Its default is 6DP, so that ~2cm (!!)
But to tolerate this we would need to have `decimal=1` .  
0.1 degrees ~1700m does seem rather large.

Reply:
Then again... the plots are rather different too...
![regrid_ProjectedUnstructured](https://user-images.githubusercontent.com/2051656/134166272-4cfffe41-4b05-4ea5-affd-ec5c55f7fac7.png)
The left plot is `cartopy` 0.19.0.post1, and the right plot is `cartopy` 0.20.0

The six subplots within each plot are the six `level`s `[0..5]` in the timeseries, where `level[0]` is the top-left and `level[5]` is bottom-right.

Apologies, forgot to add a colorbar, oops.

There is some similar structure there, but the differences are (to me) significant.

At this point, I can't confirm what's right. Any clues @pp-mo ?

Reply:
![image](https://user-images.githubusercontent.com/2051656/134256845-e261d040-1d7f-4b87-8b6b-5413c8939bbf.png)

Reply:
I don't know about the plot itself, but I see the test uses `Gnomonic` (which is WGS84 by default) and the name says UK (and seems to be the OSGB region). Could it be similar to SciTools/cartopy#1880?

Reply:
What do the maps look like from the previous two tests?  Presumably these three methods should give something fairly similar.

Reply:
> An actual change equivalent to 100's of metres does seem rather large though. Maybe the default ellipsoid for this changed ?
> I **_would_** have suggested that we should apply a more sensible tolerance to AssertArrayEqual.
> Its default is 6DP, so that ~2cm (!!)
> But to tolerate this we would need to have `decimal=1` .
> 0.1 degrees ~1700m does seem rather large.

@pp-mo Looking at the test in more detail, I don't think those values are longitude degrees, so my comparison to m makes no sense. ignore me!

Reply:
>  don't know about the plot itself, but I see the test uses `Gnomonic` (which is WGS84 by default) and the name says UK (and seems to be the OSGB region). Could it be similar to [SciTools/cartopy#1880](https://github.com/SciTools/cartopy/issues/1880)?

Awrsome, thanks @QuLogic, I'll dig a bit to see if these are related... the issue seems to be in the same space :+1:

Reply:
Note that the lower four-ish latitude rows match...
![image](https://user-images.githubusercontent.com/2051656/134326406-8419b7a0-28fc-4184-8ccd-d92bdb7d79fe.png)
Don't know how that matters wrt this issue

Reply:
My only pragmatic suggestion here is to create an issue to investigate this behaviour separately as an ongoing concern and bank this PR as is (unless there are further review actions to service).

What do we think?

Reply:
> @jamesp  Looking at the test in more detail, I don't think those values are longitude degrees,

No, on closer inspection they are just temperature values.
The 6 different levels have wildly different ranges, some very small, but the points we are testing are in level#3 which has about 1.0 deg of value range. 
I think the purpose of this test is just a backstop in case checking results stats doesn't catch a problem, like the odd missing point turning up. So I would happily increase the tolerance, just check the values to within a sensible ballpark.

I would replace the numbers with the new ones (as-is for the new ones), but also broaden the tolerance **_and_** check the masks match.
Which can be done by changing `self.assertArrayAlmostEqual(...)` to `self.assertMaskedArrayAlmostEqual(decimal=2)`
( `decimal=2` because 0.01 deg of wiggle room brings the 'old' values into range of the 'new' ones. )

Reply:
With that, **I would be ok to merge**, unless anyone else is objecting ??

Reply:
Since this is nearest neighbour regridding, the actual values shouldn't change.  But the _location_ of the values can.  In this case it looks like the points that previously were only mapped onto the bottom 4 rows are now mapped to a wider range of latitudes - seems like a genuine behaviour change.  So I'm not sure that increasing the tolerance makes sense in this context.

Reply:
> nearest neighbour regridding

Ooops.  I hadn't actually spotted that.
In that case, I guess it must be an 'unlucky' example, where the target grid is awkwardly placed.
I don't know if we can fix that somehow ?

Reply:
> actual values shouldn't change. But the location of the values can.

how did this manage to match on the array stats, I wonder ?

Reply:
I've rebased, as this PR went stale...

I'm still in favour of creating an issue to cover this, if that seems like a pragmatic compromise. Then we can dedicate some time to digging a bit further to understand the root cause of the issue (if possible), but bank this PR and move on.

Reply:
>I'm still in favour of creating an issue to cover this, if that seems like a pragmatic compromise. Then we can dedicate some time to digging a bit further to understand the root cause of the issue (if possible), but bank this PR and move on.

If we're honest, that's effectively saying we agree this change isn't a problem, and _maybe_ we'll get round to working out what caused it at some point in the future. I'm fine with that if everyone else is.

However: if this is agreed to be a real concern then I'm very uncomfortable merging into `main`. If it's a real concern and we don't have time now then we should put a maximum pin on Cartopy, no matter how horrible that prospect is.

Reply:
Does the fact that this is in `experimental` influence how much we should worry about this?

Reply:
@rcomer Nice point, it's experimental... I'm not too precious about it.

Also, the fact that this issue is isolated in one test and not an issue peppered throughout Iris is somewhat comforting.